### PR TITLE
fix(testing): use real XHR implementation in the test injector by def…

### PR DESCRIPTION
**Pull request by juliemr**
…ault

See #4539

### CODE REVIEW

**Comment by IgorMinar** [comment]
can you throw in a test that verifies that test injector binds XHR to XHRImpl? Just inject it and then do instance of check

**Comment by juliemr** [comment]
Added test.

**Comment by juliemr** [comment]
Causing issues with dart tests. I'm investigating.

**Comment by juliemr** [comment]
Possible solutions
 - bind XHR to XHRImpl only for `testing.js`, not `testing_internal.js`. I'm not thrilled with this because it introduces an odd difference between the two, but it might be the simplest solution.

 - Why do we need to run the .server.spec.dart tests? There are only a few, can they just run with browser? (more test runners => more confusion)

**Comment by juliemr** [comment]
Implemented a third solution - the test injector now provides an implementation of XHR based on `DOM.getXHR`. The DOM adapter is responsible for returning the appropriate implementation.

**Comment by juliemr** [comment]
The plot thickens. We're having errors in the benchmark tests now, because `benchmarks_external` uses the DOM adapter but does not set up reflection information. Since the DOM adapter now uses XHRImpl, which is `@Injectable`, it needs reflection information somehow.

**Comment by juliemr** [comment]
One more status update: Pending a weird failure: https://travis-ci.org/angular/angular/jobs/85429891#L535

Cannot repro locally. cc @jeffbcross @robwormald for any ideas?

**Comment by IgorMinar** [comment]
Looks like we are pulling in individual rxjs files rather than the bundle.
@jeffbcross @robwormald any idea why?

**Comment by jeffbcross** [comment]
@IgorMinar this has been the case since @vsavkin and I updated the rxjs version, because the only bundle published by rxjs right now includes all combinators (more bytes than we want).

This issue seems to be the occasionally-surfacing SystemJS loader problem where it hits some kind of limit of how many files it can load.

**Comment by jeffbcross** [comment]
I will try to update the systemjs loader and submit a PR to see if it fixes the issue. Our current version is 0.18.10, latest release is 0.19.4.

Otherwise, I don't have any quick fixes. And it's (most likely) just coincidence that the rxjs modules are the ones showing up in the logs. Historically, it has just shown the file that is trying to be loaded by the time SystemJS loader can't load files anymore.

**Comment by jeffbcross** [comment]
Though I guess the "number" of files being loaded isn't the problem in this case, since angular2 is being loaded from the bundle. I'm not sure why the examples aren't loading the rxjs that's included in the bundle :/

**Comment by jeffbcross** [comment]
Oh, except this in benchmarks, which I think don't load bundles... (pardon my rambling, I'm sleep-deprived).

**Comment by jeffbcross** [comment]
Same result after updating SystemJS loader, just with different resources failing to be loaded: https://travis-ci.org/angular/angular/jobs/85472666

**Comment by jeffbcross** [comment]
@juliemr I can help troubleshoot this tomorrow afternoon, though I'm not sure the right solution. @IgorMinar do you think the right solution would be to change our benchmarks to also use bundles?

**Comment by IgorMinar** [comment]
I've stumbled upon a similar looking issue in when updating cli, I had to define a `@reactivex` package to fix it ([see this diff](https://github.com/angular/angular-cli/pull/37/files#diff-2da5ea1f3575eb57467aeddfcca29508R20))

my guess is that more people will hit this issue when trying to load our system bundle and use `defaultJSExtensions: true` option, which sucks. I'm thinking that our recommendation should be to avoid `defaultJSExtensions` and instead load all the tsc compiled application code via a package with config `app: {defaultExtension: 'js'}`.

I'll debug why we this is failing tomorrow. From looking at the config, it should pull rx from the angular2.dev.js bundle.

**Comment by IgorMinar** [comment]
I just realized that the error is from external benchmarks, which don't load the bundle, but shouldn't load angular either.

Now I see what's going on. benchmark_utils depend on DOM adapter, which now (in this PR) depends on XHR, which depends on facade/async, which pulls in RXJS and then BAM!

let's discuss what to do with this tomorrow.

cc: @tbosch

**Comment by IgorMinar** [comment]
there are still some left over `examples` paths somewhere. e2e tests are failing because they are trying to fetch `http://localhost:8001/examples/src/async/index.html`. Can you investigate further? Also please rebase just in case your PR gets stale in the meantime.

**Comment by juliemr** [comment]
Our solution is to separate the promise facades from `facade/async` into their own `facade/promise`. XHRImpl will use only `facade/promise`, so it won't bring in rxjs stuff.

waiting for Travis to go green, then this should finally be ready!

File: modules/angular2/src/core/facade/promise.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    var reject;
+
+    var p = new Promise(function(res, rej) {
+      resolve = res;
+      reject = rej;
+    });
+
+    return {promise: p, resolve: resolve, reject: reject};
+  }
+}
```
Comment (IgorMinar) [inline modules/angular2/src/core/facade/promise.ts]: new line

File: tools/broccoli/html-replace/SCRIPTS_benchmarks_external.html [hunk 2]
```diff
@@ -12,3 +12,4 @@
   var filename = '@@PATH/@@FILENAME';
   System.import(filename).then(function(m) { m.main(); }, console.error.bind(console));
 </script>
+
```
Comment (IgorMinar) [inline tools/broccoli/html-replace/SCRIPTS_benchmarks_external.html]: not needed

File: modules/angular2/src/core/facade/async.ts [hunk 3]
```diff
@@ -1,60 +1,11 @@
 import {global, isPresent} from 'angular2/src/core/facade/lang';
+// We make sure promises are in a separate file so that we can use promises
+// without depending on rxjs.
```
Comment (IgorMinar) [inline modules/angular2/src/core/facade/async.ts]: add "e.g. in benchmarks_external"

File: modules/angular2/src/core/facade/promise.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
+// Promises are put into their own facade file so that they can be used without
+// introducing a dependency on rxjs. They are re-exported through facade/async.
```
Comment (IgorMinar) [inline modules/angular2/src/core/facade/promise.ts]: same as above

File: modules/angular2/src/core/facade/promise.ts (hunk 1 above)
Comment (juliemr) [inline modules/angular2/src/core/facade/promise.ts]: done.

File: tools/broccoli/html-replace/SCRIPTS_benchmarks_external.html (hunk 2 above)
Comment (juliemr) [inline tools/broccoli/html-replace/SCRIPTS_benchmarks_external.html]: done.

**Comment by juliemr** [comment]
Merged as d7ab5d44a586856e488934fdf3847abc89c3525c

**Pull request closed without merging**